A77-xds-rate-limiting-rlqs.md

```diff
@@ -0,0 +1,618 @@
+
+To prevent that, we'll introduce an allow-list to the bootstrap file introduced
+in [gRFC A27]. This allow-list will be a map from a fully-qualified server
+target URI to an object containing channel credentials to use for that target.
+
+```json5
+// The allowlist of Control Planes allowed to be configured via xDS.
+"allowed_grpc_services": {
+  // The key is fully-qualified server URI.
+  "dns:///xds.example.org:443": {
+    // The value is an object containing "channel_creds".
+    "channel_creds": [
+      // The format is identical to xds_servers.channel_creds.
+      {
+        "type": "string containing channel cred type",
+        "config": "optional JSON object containing config for the type"
+      }
+    ]
+  }
+}
+```
+
+When xDS Control Plane configures connection to another control plane
+via [`GrpcService.GoogleGrpc`] message, we'll inspect the allow-list for the
+matching target URI.
+
+1. If target URI is not present, we don't create the connection to the requested
+   Control Plane, and NACK the xDS resource.
+2. If target URI is present, we create the connection to the requested Control
+   Plane using the channel credentials provided in the bootstrap file. Transport
+   security configuration provided by the TD is ignored.
+
+> [!IMPORTANT]
+> This solution is not specific to RLQS, and should be used with any
+> other Control Planes configured via [`GrpcService`] message.
+
+#### Unified Matcher API
+
+RPCs will be matched into buckets using [Unified Matcher API] — an adaptable
+framework that can be used in any xDS component that needs matching features.
+
+Envoy provides two syntactically equivalent Unified Matcher
+definitions: [`envoy.config.common.matcher.v3.Matcher`](https://github.com/envoyproxy/envoy/blob/e3da7ebb16ad01c2ac7662758a75dba5cdc024ce/api/envoy/config/common/matcher/v3/matcher.proto)
+and [`xds.type.matcher.v3.Matcher`](https://github.com/cncf/xds/blob/b4127c9b8d78b77423fd25169f05b7476b6ea932/xds/type/matcher/v3/matcher.proto).
+We will only support the latter, which is the preferred version for all new APIs
+using Unified Matcher.
+
+For RLQS, Unified Matcher tree will be provided in the filter config. Evaluating
+the tree against RPC metadata will yield `RateLimitQuotaBucketSettings`, which
+contains the information needed to associate the RPC with `bucket_id` and the
+default rate limiting configuration.
+
+In this iteration the following Unified Mather extensions will be supported:
+
+1. Inputs:
+    1. [`HttpRequestHeaderMatchInput`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/http_inputs.proto#type-matcher-v3-httprequestheadermatchinput)
+    2. [`HttpAttributesCelMatchInput`](https://www.envoyproxy.io/docs/envoy/latest/xds/type/matcher/v3/http_inputs.proto#envoy-v3-api-msg-xds-type-matcher-v3-httpattributescelmatchinput)
+2. Custom Matchers:
+    1. [`CelMatcher`](https://www.envoyproxy.io/docs/envoy/latest/xds/type/matcher/v3/cel.proto.html)
```

Comment:
Doesn't this matcher require using `HttpAttributesCelMatchInput` as the input?  If so, and if this is the only matcher we're going to support, then it doesn't really make sense to say that we're going to support `HttpRequestHeaderMatchInput`, since there's no possible way to use it.

We should probably also support any common matchers that are expected to be used with `HttpRequestHeaderMatchInput`.

Reply:
`HttpRequestHeaderMatchInput` can be used with the built-in `valtype.matcher.v3.StringMatcher value_match`. I didn't specify it's explicitly because it's not an extension, as the line above implies:

> In this iteration the following Unified Mather extensions will be supported: